### PR TITLE
Support for authentication and authorization

### LINKED ISSUE #79: Validate incoming bearer token is valid Graph token for microsoft.com tenant
This allows us to protect the sample API. Related to #75

### PR DESCRIPTION
**Pull request by irvinesunday**
This PR closes #79 #80 #81 #75 #90 

Changes proposed:
- Adds support for AAD bearer tokens validation.
- Provides utility service that validates users against custom sample queries authorization policies. User claims against sample queries are stored in a json file on the root directory. 
- Adds tests to validate this utility service.
- Refactors out some old code to improve maintainability and create uniformity across the code base.
- Adds further tests to make sure code refactoring doesn't break existing implementation.

### CODE REVIEW

File: GraphExplorerSamplesService/Models/Permission.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+﻿using Newtonsoft.Json;
+using static GraphExplorerSamplesService.Models.SampleQueryModel;
```
Comment (andrueastman) [inline GraphExplorerSamplesService/Models/Permission.cs]: Remove unused using statement

File: GraphWebApi/Controllers/GraphExplorerSamplesController.cs [hunk 2]
```diff
@@ -8,32 +9,33 @@
...
+            _queriesFilePathSource = configuration["SampleQueriesFilePathName"]; // Sets the path of the sample queries JSON file
+            _policiesFilePathSource = configuration["SampleQueriesPoliciesFilePathName"]; // Sets the path of the policies file
         }
 
         // Gets the list of all sample queries
         [Route("api/[controller]")]
         [Produces("application/json")]
         [HttpGet]
-        public async Task<IActionResult> GetSampleQueriesList(string search)
+        public async Task<IActionResult> GetSampleQueriesListAsync(string search)
```
Comment (andrueastman) [inline GraphWebApi/Controllers/GraphExplorerSamplesController.cs]: 👍

File: GraphWebApi/Controllers/GraphExplorerSamplesController.cs [hunk 3]
```diff
@@ -185,12 +229,37 @@ public async Task<IActionResult> CreateSampleQuery([FromBody]SampleQueryModel sa
...
+                bool isAuthorized = SamplesPolicyService.IsUserAuthorized(policies, userPrincipalName, categoryName, HttpMethods.Delete);
+
+                if (!isAuthorized)
+                {
+                    return new JsonResult(
+                        $"{userPrincipalName} is not authorized to delete the sample query. Category: '{categoryName}'")
+                    { StatusCode = StatusCodes.Status401Unauthorized };
+                }
 
                 if (sampleQueriesList == null || sampleQueriesList.SampleQueries.Count == 0)
```
Comment (andrueastman) [inline GraphWebApi/Controllers/GraphExplorerSamplesController.cs]: nitpick: `sampleQueriesList` will never be null as `GetSampleQueriesListAsync` will return an empty list or a non empty list.

File: SamplesService.Test/SampleQueryModelShould.cs [hunk 4]
```diff
@@ -149,6 +149,5 @@ public void TrimAllLeadingAndTrailingWhiteSpacesFromDocLinkValueIfPresent()
         }
 
         #endregion
-
```
Comment (andrueastman) [inline SamplesService.Test/SampleQueryModelShould.cs]: nitpick: unnecessary diff

File: SamplesService.Test/SamplesPolicyServiceShould.cs [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+        #region Create Sample Queries Policies Template Tests
+
+        [Fact]
+        public void CreateDefaultPoliciesTemplate()
+        {
+            // Arrange and Act
+            SampleQueriesPolicies policiesTemplate = SamplesPolicyService.CreateDefaultPoliciesTemplate();
+
+            // Assert
+            Assert.NotEmpty(policiesTemplate.CategoryPolicies);
```
Comment (andrueastman) [inline SamplesService.Test/SamplesPolicyServiceShould.cs]: Does it make sense to validate the structure of the default template is as expected?

File: GraphWebApi/Controllers/GraphExplorerSamplesController.cs (hunk 3 above)
Comment (irvinesunday) [reply GraphWebApi/Controllers/GraphExplorerSamplesController.cs]: I agree.

File: GraphExplorerSamplesService/Models/Permission.cs (hunk 1 above)
Comment (irvinesunday) [reply GraphExplorerSamplesService/Models/Permission.cs]: Done

File: SamplesService.Test/SamplesPolicyServiceShould.cs (hunk 5 above)
Comment (irvinesunday) [reply SamplesService.Test/SamplesPolicyServiceShould.cs]: I agree. It makes sense validating that the policy templates created follow a certain specified structure.
